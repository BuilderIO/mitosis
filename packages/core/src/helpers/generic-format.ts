const preSpaceRegex = /^\s*/g;

const DEFAULT_INDENT_SPACES = 2;

/**
 * Generic formatter for languages prettier doesn't support, like Swift
 *
 * Not super sophisticated, but much better than nothing
 */
export const format = (str: string, indentSpaces = DEFAULT_INDENT_SPACES) => {
  let currentIndent = 0;
  const lines = str.split('\n');
  lines.forEach((item, index) => {
    item = item.trimEnd();

    if (!item) {
      lines[index] = '';
      return;
    }

    lines[index] = item.replace(preSpaceRegex, ' '.repeat(currentIndent * indentSpaces));

    const nextLine = lines[index + 1];
    if (!nextLine) {
      return;
    }

    if (nextLine.match(/^\s*[})][,;]?\s*$/)) {
      currentIndent--;
    } else if (item.match(/([({]| in)$/)) {
      currentIndent++;
    }

    currentIndent = Math.max(currentIndent, 0);
  });
  return lines.join('\n').replace(/\n{3,}/g, '\n\n');
};
