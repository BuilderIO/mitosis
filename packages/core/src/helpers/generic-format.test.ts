import { format } from './generic-format';

test('Can generic format sloppy Swift code', () => {
  const code = `
  import SwiftUI
  import JavaScriptCore
  
  struct MyComponent: View {
    
      @State private var updateIndex = 0
      private var context = JSContext()
  
      func eval(expression: String) -> JSValue! {
        return context?.evaluateScript(expression)
      }
  
      init() {
        let jsSource = """
            const state = { name: "Steve" };
  
        """
        context?.evaluateScript(jsSource)
      }
    
  
    var body: some View {
      VStack {
        Text(String(updateIndex)).hidden()
        VStack(){
        Foo(
          bar: baz
          )
  TextField(){}
  Text("Hello")
  Text(eval(expression: """state.name"""))
  Text("! I can run in React, Vue, Solid, or Liquid!")}
      }
    }
  }
  `;

  expect(format(code)).toMatchSnapshot();
});
