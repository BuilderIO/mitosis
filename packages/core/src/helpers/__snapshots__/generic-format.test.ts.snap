// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Can generic format sloppy Swift code 1`] = `
"
import SwiftUI
import JavaScriptCore

struct MyComponent: View {

  @State private var updateIndex = 0
  private var context = JSContext()

  func eval(expression: String) -> JSValue! {
    return context?.evaluateScript(expression)
  }

  init() {
    let jsSource = \\"\\"\\"
    const state = { name: \\"Steve\\" };

    \\"\\"\\"
    context?.evaluateScript(jsSource)
  }

  var body: some View {
    VStack {
      Text(String(updateIndex)).hidden()
      VStack(){
        Foo(
          bar: baz
        )
        TextField(){}
        Text(\\"Hello\\")
        Text(eval(expression: \\"\\"\\"state.name\\"\\"\\"))
        Text(\\"! I can run in React, Vue, Solid, or Liquid!\\")}
      }
    }
  }
"
`;
